### PR TITLE
chore(datadog_metrics sink): support and migrate to the `v2` series API endpoint

### PR DESCRIPTION
**Pull request by neuronull**
This change adds support for the `datadog_metrics` sink to encode Series metrics to the Datadog Series `v2` API endpoint and changes the encoder to use that endpoint by default. Tests were updated to cover the case of the v2 endpoint, which encodes metrics as ProtoBuf, similar to Sketches. This PR lightly refactored the calling logic for the incremental encoding of Sketches, because the Series implementation was much the same.

The `v1`  endpoint will be supported for another full release, and users are able to use the v1 endpoint by setting an environment variable `VECTOR_TEMP_USE_DD_METRICS_SERIES_V1_API` , which will be removed when the `v1` endpoint support is removed.

In addition to the unit and integration test coverage that was added, this change was tested manually by running the Agent locally with the `datadog_agent` source and the `datadog_metrics` sink.

### CODE REVIEW

File: src/sinks/datadog/metrics/config.rs [hunk 1]
```diff
@@ -42,27 +42,58 @@ impl SinkBatchSettings for DatadogMetricsDefaultBatchSettings {
...
+        }
+    }
+}
+
+impl Default for SeriesApiVersion {
+    fn default() -> Self {
+        Self::V2
+    }
+}
+
```
Comment (tobz) [inline src/sinks/datadog/metrics/config.rs:70]: What's the goal with all the configurable stuff here? Seems like the intent to move wholesale to the v2 API, since there's nothing exposed to allow staying on/switching back to the v1 API, at least from the user's perspective.

Comment (neuronull) [reply src/sinks/datadog/metrics/config.rs:70]: Yeah I was going to call this out when I finalized the PR, but I recall chatting with @jszwedko about this (the history of the convo is gone though) . I don't remember the details but I believe the conclusion was we wanted to keep the v1 logic if it wasn't too difficult to do so. We had discussed allowing the api version to be configurable somehow, but felt that wasn't necessary.

So I _think_ it might boil down to us wanting to be able to flip a switch to go back to V1 if there were issues with our rollout of V2.

Another idea I had thrown out was having a non publicized backdoor essentially, that would allow a user to configure to use the v1. Its possible we could do that with the `endpoint` configuration I think. We'd just have to have some uglier logic in the parsing of that for this sink, since it isn't supposed to specify the `path`, just the host/port.

TLDR ... Its open for discussion 😆

Comment (jszwedko) [reply src/sinks/datadog/metrics/config.rs:70]: Yeah, I see including the fallback as a risk mitigation measure so that if it does break things, users have a workaround until we fix it. We can weigh the risk against the level of effort to maintain the v1 implementation though.

Comment (neuronull) [reply src/sinks/datadog/metrics/config.rs:70]: > users have a workaround until we fix it

In it's current state, users would have to edit the code and recompile in order to take advantage of that. But, my suggestion here:

> Another idea I had thrown out was having a non publicized backdoor essentially, that would allow a user to configure to use the v1. Its possible we could do that with the endpoint configuration I think. We'd just have to have some uglier logic in the parsing of that for this sink, since it isn't supposed to specify the path, just the host/port.

, would allow us to give users a workaround that would just entail editing their config. If we did this, could tag it with a TODO and followup GH issue to remove the backdoor and the v1 support altogether, once it had enough soak time.

File: src/sinks/datadog/metrics/encoder.rs [hunk 2]
```diff
@@ -882,15 +1013,16 @@ mod tests {
     };
 
     use super::{
```
Comment (dsmith3197) [inline src/sinks/datadog/metrics/encoder.rs:1015]: 💬 Is there value in adding corresponding tests here for the Series V2 logic? From my understanding, these test currently only exercise the Sketch and V1 Series logic.

File: src/sinks/datadog/metrics/integration_tests.rs [hunk 3]
```diff
@@ -104,16 +126,105 @@ async fn smoke() {
...
+
+    let series = payload.series;
+
+    assert!(!series.is_empty());
+
+    // check metrics are sorted by name, which helps HTTP compression
+    let metric_names: Vec<String> = series.iter().map(|serie| serie.metric.clone()).collect();
+    let mut sorted_names = metric_names.clone();
+    sorted_names.sort();
+    assert_eq!(metric_names, sorted_names);
```
Comment (dsmith3197) [inline src/sinks/datadog/metrics/integration_tests.rs:166]: 💭 Is this true for proto payloads as well? Currently, we have two reasons for why we sort the metrics in a payload: (1) for improved compression and (2) to aggregate counters. In the future, when we revamp the aggregation logic, the second reason will be obsolete. And if this doesn't hold true for proto payloads either, then we can remove the need to sort payloads altogether when we remove support for the V1 endpoint.

File: src/sinks/datadog/metrics/integration_tests.rs [hunk 4]
```diff
@@ -104,16 +126,105 @@ async fn smoke() {
...
+
+        // tags
+        assert_eq!(serie.tags.len(), 1);
+        assert_eq!(serie.tags.first().unwrap(), "cool_tag_name:i_know_right");
+
+        // unit
+        assert!(serie.unit.is_empty());
+
+        // interval
+        assert_eq!(serie.interval, 0);
```
Comment (dsmith3197) [inline src/sinks/datadog/metrics/integration_tests.rs:199]: 💭 It would be good to exercise more code paths in our tests. Specifically, testing (1) a counter with an interval specified, (2) a gauge, and (3) a set.

**Review by dsmith3197 — COMMENTED** [review]
Looks good overall. Just some small comments.

File: src/sinks/datadog/metrics/encoder.rs [hunk 5]
```diff
@@ -394,6 +440,22 @@ impl DatadogMetricsEncoder {
...
+    maybe_pass_through: Option<&DatadogMetricOriginMetadata>,
+    maybe_source_type: Option<&'static str>,
+    origin_product_value: u32,
+) -> Option<ddmetric_proto::Metadata> {
+    generate_origin_metadata(maybe_pass_through, maybe_source_type, origin_product_value).map(
+        |origin| ddmetric_proto::Metadata {
+            origin: Some(ddmetric_proto::Origin {
+                origin_product: origin.product().expect("OriginProduct should be set"),
+                origin_category: origin.category().expect("OriginCategory should be set"),
+                origin_service: origin.service().expect("OriginService should be set"),
```
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sinks/datadog/metrics/encoder.rs:453]: At a high level I agree that its safer to default and log.

But here specifically, if the `generate_origin_metadata()` gave us a Some(origin) , then its a bug if those aren't set. Which actually makes me realize something- in the beginning of the Origin PR I created those functions to set each individual value. But by the end of the PR, we changed the design such that if we set any of the Origin values, we set  all of them. Therefore I feel the functions to set each value should be replaced with a single function to set all. That leaves us less room to accidentally not set one.

I suppose the concern still remains though, so I guess it would still be reasonable to log a warning.

However, all of that said, since this PR hasn't modified any of that, I'd treat this as a tertiary concern and would opt to change that in a separate PR.

WDYT?

File: src/sinks/datadog/metrics/config.rs (hunk 1 above)
Comment (neuronull) [reply src/sinks/datadog/metrics/config.rs:70]: I can get behind a temporary env var.

> Yeah, I see including the fallback as a risk mitigation measure so that if it does break things, users have a workaround until we fix it. We can weigh the risk against the level of effort to maintain the v1 implementation though.

To clarify- IMO I think we would only want to maintain the v1 implementation for a full release, maybe two. But after its clear there aren't issues, we'd remove it. In my view the maintenance burden of supporting the v1 implementation for a release or two should be very small if not zero.

File: src/sinks/datadog/metrics/encoder.rs (hunk 2 above)
Comment (neuronull) [reply src/sinks/datadog/metrics/encoder.rs:1015]: There are a couple that are clear cases for extending to the v2, such as the `incorrect_metric_for_endpoint_causes_error` and `encode_single_series_metric_with_default_limits`.

However its less clear to me the value of extending the existing test cases that cover the payload size limits etc. Curious if @tobz  has opinions there.

(Side note ofc, the integration test has been extended to cover the v2 case).

File: src/sinks/datadog/metrics/integration_tests.rs (hunk 3 above)
Comment (neuronull) [reply src/sinks/datadog/metrics/integration_tests.rs:166]: Mmm yeah, naively I would think that the sort by name wouldn't matter for the proto case. I kept the "check" just to be consistent but I think it makes sense to remove it from the v2 checks, and then if/when we remove the v1 support, we'll remove the v1 test case here. WDYT?

File: src/sinks/datadog/metrics/encoder.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsmith3197) [reply src/sinks/datadog/metrics/encoder.rs:453]: I'm happy for this to be done in a separate PR 👍🏻

File: src/sinks/datadog/metrics/integration_tests.rs (hunk 3 above)
Comment (dsmith3197) [reply src/sinks/datadog/metrics/integration_tests.rs:166]: My hunch is that the performance gains still apply for the proto case because it also use zlib compression.

File: src/sinks/datadog/metrics/encoder.rs (hunk 2 above)
Comment (neuronull) [reply src/sinks/datadog/metrics/encoder.rs:1015]: btw, I already started extending these, even more than the ones I listed

File: src/sinks/datadog/metrics/config.rs (hunk 1 above)
Comment (neuronull) [reply src/sinks/datadog/metrics/config.rs]: I added the env var in the latest changes and made code comments regarding the temporary support of V1. Since this isn't a user-facing deprecation/change, I dont think we need to follow the formal DEPRECATION policy if I'm not mistaken

Comment (neuronull) [reply src/sinks/datadog/metrics/config.rs]: (Which I validated works by injecting it in the vdev env vars and seeing the debug logs from the integration tests)

File: src/sinks/datadog/metrics/encoder.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sinks/datadog/metrics/encoder.rs:453]: 👍 
created https://github.com/vectordotdev/vector/pull/18788

File: src/sinks/datadog/metrics/config.rs (hunk 1 above)
Comment (neuronull) [reply src/sinks/datadog/metrics/config.rs]: > Since this isn't a user-facing deprecation/change, I dont think we need to follow the formal DEPRECATION policy if I'm not mistaken

As discussed, I noted this in the DEPRECATIONS file but am holding off on a mention to the upgrade guide as its not a user facing change and is lowish risk

**Review by dsmith3197 — APPROVED** [review]
